File: compiler/bench/bitscan.fs
\ expected: 9999985
\ Bit scan forward - find lowest set bit position

: bsf ( n -- pos )
  dup 0= if drop -1 exit then
  0 swap
  begin dup 1 and 0= while
    1 rshift swap 1+ swap
  repeat drop ;

: main
  0
  10000000 1 do
    i bsf +
  loop
  . cr ;

File: compiler/bench/bitscan.c
// expected: 9999985
// Bit scan forward - find lowest set bit position
#include <stdio.h>
#include <stdint.h>

int bsf(uint32_t n) {
    if (n == 0) return -1;
    int pos = 0;
    while ((n & 1) == 0) {
        n >>= 1;
        pos++;
    }
    return pos;
}

int main(void) {
    long sum = 0;
    for (int i = 1; i < 10000000; i++) {
        sum += bsf(i);
    }
    printf("%ld\n", sum);
    return 0;
}
